### PR TITLE
mesa: split out the Darwin build into a separate expression

### PR DESCRIPTION
**Pull request by K900**
The way we build Mesa for Linux (and potentially others) is very, very different from the way we build Mesa on Darwin. In short, on Linux we generally want to build as much of it as possible, while on Darwin we want to build just enough of it to get GLX for legacy X applications.

So, split out the Darwin build into its own expression entirely, and give it just enough inputs and just enough feature flags to actually do the one thing we want it to do, and nothing else.

Then use the newly gained flexibility to remove most of the platform conditionals from the Linux expression.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -23285,10 +23285,12 @@ with pkgs;
   # Default libGLU
   libGLU = mesa_glu;
 
-  mesa = darwin.apple_sdk_11_0.callPackage ../development/libraries/mesa {
-    inherit (darwin.apple_sdk_11_0.frameworks) OpenGL;
```
Comment (alyssais) [inline pkgs/top-level/all-packages.nix:23289]: OpenGL no longer required?

Comment (K900) [reply pkgs/top-level/all-packages.nix:23289]: I don't think it ever was, actually - it's only ever dlopened from a fixed path.

**Comment by K900** [comment]
Re-added split outputs to the Darwin expression to make libGL stubs build again.

File: pkgs/development/libraries/mesa/default.nix [hunk 2]
```diff
@@ -204,7 +173,7 @@ self = stdenv.mkDerivation {
     "out" "dev" "drivers"
   ] ++ lib.optionals enableOSMesa [
     "osmesa"
-  ] ++ lib.optionals stdenv.isLinux [
+  ] ++ [
     "driversdev"
```
Comment (philiptaron) [inline pkgs/development/libraries/mesa/default.nix:177]: I'd move this up to the unilateral section so it looks neater, even thought it causes a rebuild.

File: pkgs/development/libraries/mesa/default.nix [hunk 3]
```diff
@@ -100,28 +93,19 @@
...
 
 # When updating this package, please verify at least these build (assuming x86_64-linux):
 # nix build .#mesa .#pkgsi686Linux.mesa .#pkgsCross.aarch64-multiplatform.mesa .#pkgsMusl.mesa
 
 let
-  version = "24.1.1";
-  hash = "sha256-ADiCbG9+iNkLTOb3GRkvpYyn3t9O3KoRdM972SDvieo=";
-
-  # Release calendar: https://www.mesa3d.org/release-calendar.html
-  # Release frequency: https://www.mesa3d.org/releasing.html#schedule
```
Comment (philiptaron) [inline pkgs/development/libraries/mesa/default.nix:120]: These want to move over to common.nix

**Review by philiptaron — APPROVED** [review]
Read through the Nix code carefully and built on x86_64-darwin x86_64-linux. Couple of in-no-way-blocking comments.

[ofBorg is unhappy](https://gist.github.com/GrahamcOfBorg/ba39f905102286c86bfa9cb02119ca46), though, due to `pkgs/development/libraries/libva/default.nix` on Darwin. (It's marked as unsupported, but it doesn't eval with this PR.)

File: pkgs/top-level/all-packages.nix [hunk 4]
```diff
@@ -23285,10 +23285,12 @@ with pkgs;
...
-    inherit (darwin.apple_sdk_11_0.libs) Xplugin;
-  };
+  mesa = if stdenv.isDarwin
+    then darwin.apple_sdk_11_0.callPackage ../development/libraries/mesa/darwin.nix {
```
[the remark this answers is not in the document]
Comment (K900) [reply pkgs/top-level/all-packages.nix:23289]: This is the newest one we have.

File: pkgs/development/libraries/mesa/default.nix (hunk 3 above)
Comment (K900) [reply pkgs/development/libraries/mesa/default.nix:120]: They're really not relevant information.

File: pkgs/development/libraries/mesa/default.nix (hunk 2 above)
Comment (K900) [reply pkgs/development/libraries/mesa/default.nix:177]: I'll do more cleanups in a follow-up PR to staging.

**Review by philiptaron — APPROVED** [review]

**Comment by vcunat** [comment]
@ofBorg eval

**Comment by vcunat** [comment]
ofBorg is still broken.  An assertion on the right place would do it, I think.  Or maybe there's some better way.

**Comment by K900** [comment]
Retargeted to next staging, current staging is unblocked by horrible hack.

**Comment by philiptaron** [comment]
On `x86_64-darwin`, I went through each of these derivations and tried to build them directly. The only one that failed to build also failed to build on master: `nix-build -A python3Packages.pyopencl`.

I did not try to build on x86_64-linux yet.

**Review by emilazy — APPROVED** [review]
I hacked up glxgears to build on macOS and got a blazing 12445.432 FPS on XQuartz with this PR applied. LGTM

**Comment by Ninlives** [comment]
This PR breaks mesa build on riscv, as the valgrind is currently not supported on riscv, and the option for removing it from the buildInputs is deleted in this PR.

**Pull request merged**